### PR TITLE
Initial version of pre-commit hooks

### PR DESCRIPTION
**Pull request by krish2718**
This has standard checks that are used in BW org across repositories.

### CODE REVIEW

File: README.md
Comment (bwt-sloanj) [inline README.md]: This isn't referenced anywhere.

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: This approach can be updated for Python 3.10+, i.e. use `dict` directly in type hints etc. In some cases `whatever | None` might also make more sense than `Optional[whatever]`.

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: Use `importlib.resources`.

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: If we say that we are on Python 3.10+ then this can be removed and its usage replaced with the new `Path` method.

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: Maybe this could just be `return commit_msg_file.read_text()`

(I like `pathlib`...)

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: f-string?

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: Would `check_output()` be appropriate here?

File: bw_check_commit_hook/data/checkpatch.yaml
Comment (bwt-sloanj) [inline bw_check_commit_hook/data/checkpatch.yaml]: This syntax doesn't seem ideal - could it at least become `"."` to make it a little clearer?

File: setup.py
Comment (bwt-sloanj) [inline setup.py]: In general we want to use BWT instead of BW, and the repo/package should be `check-commit-hook` (`checkpatch-hook` would be better really...). This applies everywhere, including directory names, i.e. change the dir name to `check_commit_hook`.

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: I think I would prefer the `sys.exit(1)` to occur within this function instead of outside it. The `except` clause could instead catch the specific exceptions that are expected to occur for whichever scenarios, and the `if key not in config` clause could throw a `ValueError` or `RuntimeError` to be caught in that `except` clause.

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: Could this determination of what config to apply within a directory be refactored into a separate function?

This code also just seems a bit strange, for example it is relying on very similar key names where one is uppercase and the other lowercase.

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [inline bw_check_commit_hook/__init__.py]: Could you please apply `black` to this? Adding it as a hook to be run on this repo itself would be good.

File: bw_check_commit_hook/data/spelling.txt
Comment (bwt-sloanj) [inline bw_check_commit_hook/data/spelling.txt:1]: Is this file needed now?

**Review by bwt-sloanj — CHANGES_REQUESTED** [review]
Since this is being adapted from the internal `check_commit.py` it is worth bearing in mind that this hook is currently only implementing the `checkpatch` part of that script - the python function `check_commit_message()` isn't part of it (probably something I cut out). It therefore needs to be considered if this hook needs to be expanded to also do those checks, or if another hook could do those, or if they aren't wanted for this.

Also, I appreciate that in reviewing this that I am in effect requesting changes to code that you didn't write yourself, and in some cases that I wrote, but what I gave you wasn't complete and we have also migrated Python versions since then.

**Comment by krish2718** [comment]
>  It therefore needs to be considered if this hook needs to be expanded to also do those checks, or if another hook could do those, or if they aren't wanted for this.

Well, the per-commit checks are too specific to BWT, so, IMHO, we should drop that for now, and as a first cut stick with checkpatch + spellchecks (codespell). And also, we can just use `gitlint` instead of custom checking, that can be another hook in this repo added later.

File: bw_check_commit_hook/__init__.py
Comment (krish2718) [reply bw_check_commit_hook/__init__.py]: I am a bit confused, moving `sys.exit(1)` to this function and specific exception handling seems conflicting. IIUC, you are suggesting something like this.

```
def _load_config(config_file: Path) -> dict:
    mandatory_keys = ["DIR_CONFIGS"]
    try:
        with config_file.open("r") as file:
            config = yaml.safe_load(file)
        if config is None or not isinstance(config, dict) or not config:
            raise RuntimeError("Config file is empty")
        for key in mandatory_keys:
            if key not in config:
                raise RuntimeError(f"Missing mandatory key {key} in config file")
        return config
    except FileNotFoundError:
        print(f"Config file not found: {config_file}")
        sys.exit(1)
    except yaml.YAMLError as e:
        print(f"Error loading YAML config file: {e}")
        sys.exit(1)
    except RuntimeError as e:
        sys.exit(1)
```

File: bw_check_commit_hook/data/checkpatch.yaml
Comment (krish2718) [reply bw_check_commit_hook/data/checkpatch.yaml]: I can add a magic string `all` and then handle the conversion in the code, WDYT?

File: bw_check_commit_hook/__init__.py
Comment (krish2718) [reply bw_check_commit_hook/__init__.py]: It's deprecated in 3.12+ https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to, so, better to use the abstraction to hide the version handling.

File: bw_check_commit_hook/__init__.py
Comment (krish2718) [reply bw_check_commit_hook/__init__.py]: No, we need to catch and read errors.

File: bw_check_commit_hook/data/spelling.txt
Comment (krish2718) [reply bw_check_commit_hook/data/spelling.txt:1]: Yes, this is different from codespell, without this file it will throw a warning.

File: bw_check_commit_hook/__init__.py
Comment (bwt-sloanj) [reply bw_check_commit_hook/__init__.py]: The deprecation is for supporting multiple arguments, i.e. the signature will change from taking `*others` to taking `other` as a single argument. So it will still support our use in 3.12+ and can remain.

File: .pre-commit-config.yaml
Comment (bwt-sloanj) [inline .pre-commit-config.yaml]: We use `black` for style - so the `double-quote-string-fixer` will also have to go.

File: bw_check_commit_hook/data/checkpatch.yaml
Comment (bwt-sloanj) [reply bw_check_commit_hook/data/checkpatch.yaml]: `all` makes it sound like it would apply in all sub-directories regardless of whether they have their own entries in the YAML. Maybe `default` would be better? Or maybe there's a better convention out there somewhere.

Comment (krish2718) [reply bw_check_commit_hook/data/checkpatch.yaml]: `default` works for me.

File: bw_check_commit_hook/__init__.py
Comment (krish2718) [reply bw_check_commit_hook/__init__.py]: Ah ok, thanks.

File: tests/test_config.py
Comment (bwt-sloanj) [inline tests/test_config.py:1]: It looks like the hooks haven't run on this file.

File: checkpatch_hook/data/checkpatch.yaml
Comment (bwt-sloanj) [inline checkpatch_hook/data/checkpatch.yaml]: Hmm how about `__default__` to really avoid a potential name collision and because double underscores are very Pythonic.

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py]: Could this be tweaked a bit to move the parsing of the errors into a function?

So more like `if result.returncode != 0: # parse errors`?

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py]: I think I'd prefer this to be either indicated as a kind of constant, i.e. `DEFAULT_CONFIG_FILE` in the global context, or just loaded directly into the `default=` kwarg, instead of being a local variable in this function.

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py]: At the moment the strings you are providing above when you `raise RuntimeError(...)` are being lost. I think what I was asking for earlier would probably require replacing the use of `RuntimeError` with a custom exception type that lets you retrieve the message and print that:

```python
class _ConfigException(BaseException):
    def __init__(..., msg: str):
        super().__init__(...)
        self.msg = msg
```

then you just print `e.msg` in that particular `except` block.

BTW, is printing to `sys.stderr` necessary? Maybe see if there's a convention for where some widely-used hooks handle that kinda thing.

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py]: Needs to be updated, or probably better, just say it's a special case.

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py]: To be replaced.

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py]: To be removed.

**Review by bwt-sloanj — CHANGES_REQUESTED** [review]

File: checkpatch_hook/__init__.py
Comment (krish2718) [reply checkpatch_hook/__init__.py]: The strings are not lost, as I printing `{e}`, see

```
$ checkpatch --config-file test.yaml
Error loading YAML config file: Missing mandatory key DIR_CONFIGS in test.yaml
```

> BTW, is printing to sys.stderr necessary? Maybe see if there's a convention for where some widely-used hooks handle that kinda thing.

Yes, errors are being printed to stderr in most hooks.

File: tests/test_config.py
Comment (krish2718) [reply tests/test_config.py:1]: I don't know why it didn't earlier, but I now checked and they ran.

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [reply checkpatch_hook/__init__.py]: Ah I didn't realise that printing an exception behaved like that - well that looks good to me then if you're happy too.

File: .pre-commit-hooks.yaml
Comment (bwt-sloanj) [inline .pre-commit-hooks.yaml:1]: In the examples that you've given me of use of the hooks in other repos you have entries like:

```
files: ^.*\.(c|h|cpp|hpp|cc|hh|cxx|hxx|py|sh|yaml|yml)$
```

Look at `types_or` and `types` in https://pre-commit.com/#new-hooks and see if that could be used here to move that kind of filtering into this repo. It uses `identify-cli` which is pretty neat for identifying what different files are.

Comment (krish2718) [reply .pre-commit-hooks.yaml:1]: Sure, I can use `types_or` or `types` instead of `files`, but as we had discussed earlier, we can leave this to `checkpatch.pl` to figure it out as it has the intelligence and might add checks for new filetypes in the future (or remove).

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py:150]: Could you refactor slightly to avoid the repetition of this logging statement?

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py]: Here is some sample output when running with `--verbose`:

```
DEBUG:checkpatch_hook:No changes to check in <dir1>
DEBUG:checkpatch_hook:Running checkpatch with: checkpatch.pl --strict --no-tree --emacs --terse --showfile --no-summary --color=never --no-signoff --ignore BAD_SIGN_OFF,SPDX_LICENSE_TAG,FILE_PATH_CHANGES,NOT_UNIFIED_DIFF,LINUX_VERSION_CODE,CONSTANT_COMPARISON,OPEN_ENDED_LINE,UNNECESSARY_PARENTHESES,GERRIT_CHANGE_ID,COMMIT_LOG_LONG_LINE,EMAIL_SUBJECT,GIT_COMMIT_ID --max-line-length=120
DEBUG:checkpatch_hook:No changes to check in <dir3>
DEBUG:checkpatch_hook:No changes to check in <dir4>
DEBUG:checkpatch_hook:No changes to check in <dir5>
ERROR:checkpatch_hook:<dir2>/something.c:
ERROR:checkpatch_hook:  85: checkpatch: WARNING: please, no spaces at the start of a line
```

The `DEBUG` statement with the args doesn't state which entry from `DIR_CONFIGS` it is derived from, so when multiple directories are being triggered which args apply to which dirs will be ambiguous (well it looks ambiguous with the real dir names...). Just passing in `config_dir` as a third arg and adding it to the log statement should be fine.

File: README.md
Comment (bwt-sloanj) [inline README.md]: `bw_check_commit_hook` is outdated.

File: README.md
Comment (bwt-sloanj) [inline README.md]: "all the files" is maybe misleading since it is only be selectively applied. Really the original script had quite a strange/specific way of sharing config between dirs (called COMMON but not always actually common). What do you think? Might it be better to encourage use of YAML anchors? Then `pyyaml` does the work and it is a clearer and documented feature of the config file. The user can alternatively just write out all the config options for each directory in full.

File: README.md
Comment (bwt-sloanj) [inline README.md]: This would normally be a tag - better to put a generic placeholder rather than a specific commit SHA.

File: .pre-commit-config.yaml
Comment (bwt-sloanj) [inline .pre-commit-config.yaml]: Development note?

**Review by bwt-sloanj — CHANGES_REQUESTED** [review]

File: checkpatch_hook/__init__.py
Comment (krish2718) [reply checkpatch_hook/__init__.py:150]: It's not repetition, both are different `files` vs `changes`, this adds value to the debug.

File: README.md
Comment (krish2718) [reply README.md]: That was my first implementation, but anchors weren't' extending in to a single list they show as nested lists and I didn't want to change the core logic too much, so, abandoned the approach.

File: README.md
Comment (krish2718) [reply README.md]: Sure, pre-commit shows a warning that non-specific revisions are only updated once, so, I was using the commit SHA, but yes, this should be replaced by a tag once the PR is merged.

File: .pre-commit-config.yaml
Comment (krish2718) [reply .pre-commit-config.yaml]: Yes, will remove

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [reply checkpatch_hook/__init__.py:150]: Silly me

File: README.md
Comment (bwt-sloanj) [reply README.md]: To potentially be investigated in the future.

File: README.md
Comment (krish2718) [reply README.md]: Sorry, didn't realize that this is a doc :), replaced with xxx.

File: checkpatch_hook/__init__.py
Comment (bwt-sloanj) [inline checkpatch_hook/__init__.py]:
```suggestion
    logger.debug(f"Running checkpatch in {config_dir_str} with: {' '.join(cmd)}")
```
Would be consistent with the other log message and the wording is slightly more accurate.

**Pull request merged**
